url7a45f192ca_mfullen_ping_pong_net_tgz-pJ8-ping_pong_net_client_io_TestClientNonPPNServerJ8/src/ping/pong/net/connection/io/AbstractIoConnection.java

```diff
@@ -0,0 +1,475 @@
+        this.ioTcpWriteRunnable = new IoTcpWriteRunnable<MessageType>(new RunnableEventListener()
+        {
+            @Override
+            public void onRunnableClosed(DisconnectState disconnectState)
+            {
+                LOGGER.trace("ioTcpWriteRunnable closed {}", disconnectState);
+                disconnect();
+                ioTcpWriteRunnable = null;
+            }
+        }, dataWriter, tcpSocket);
+
+        return successful;
+    }
+
+    @Override
+    public boolean isUsingCustomSerialization()
+    {
+        return usingCustomSerialization;
+    }
+
+    /**
+     *
+     */
+    protected void disconnect()
+    {
+        this.receiveQueue.add((MessageType) new DisconnectMessage());
+    }
+
+    @Override
+    public synchronized SocketAddress getSocketAddress()
+    {
+        if (this.socketAddress == null)
+        {
+            throw new NullPointerException("Socket not connected yet");
+        }
+        return this.socketAddress;
+    }
+
+    /**
+     * Fire the on socket created event for all listeners
+     */
+    protected synchronized void fireOnSocketCreated()
+    {
+        for (ConnectionEvent connectionEvent : connectionEventListeners)
+        {
+            connectionEvent.onSocketCreated();
+        }
+    }
+
+    /**
+     * Fire the on socketMessage Received event for all listeners
+     * @param message
+     */
+    protected synchronized void fireOnSocketMessageReceived(MessageType message)
+    {
+        for (ConnectionEvent<MessageType> connectionEvent : connectionEventListeners)
+        {
+            connectionEvent.onSocketReceivedMessage(message);
+        }
+    }
+
+    /**
+     *
+     */
+    protected synchronized void fireOnSocketClosed()
+    {
+        for (ConnectionEvent connectionEvent : this.connectionEventListeners)
+        {
+            connectionEvent.onSocketClosed();
+        }
+    }
+
+    /**
+     *
+     */
+    protected void startConnection()
+    {
+        if (this.canStart)
+        {
+            this.executorService.execute(ioTcpWriteRunnable);
+            this.executorService.execute(ioTcpReadRunnable);
+
+            this.startUdp();
+
+            this.connected = true;
+            LOGGER.trace("Connection started successfully.");
+        }
+        else
+        {
+            LOGGER.error("This connection cannot start because it was not initialized properly.");
+        }
+    }
+
+    @Override
+    public void run()
+    {
+        startConnection();
+
+        while (this.isConnected())
+        {
+            try
+            {
+                LOGGER.trace("({}) About to block to Take message off queue", this.getConnectionId());
+                MessageType message = this.receiveQueue.take();
+
+                LOGGER.trace("({}) Message taken to be processed ({})", this.getConnectionId(), message);
+                this.processMessage(message);
+            }
+            catch (InterruptedException ex)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)